src/linear_algebra/affine_space.lean

```diff
@@ -0,0 +1,267 @@
+variables [S : affine_space k V P]
+include S
+
+/-- The submodule spanning the differences of a (possibly empty) set
+of points. -/
+def vector_span (s : set P) : submodule k V := submodule.span k (vsub_set V s)
+
+/-- The points in the affine span of a (possibly empty) set of
+points. Use `affine_span` instead to get an `affine_subspace k V P`,
+if the set of points is known to be nonempty. -/
+def span_points (s : set P) : set P :=
+{p | ∃ p1 ∈ s, ∃ v ∈ (vector_span k V s), p = v +ᵥ p1}
+
+/-- A point in a set is in its affine span. -/
+lemma mem_span_points (p : P) (s : set P) : p ∈ s → p ∈ span_points k V s
+| hp := ⟨p, hp, 0, submodule.zero _, (zero_vadd V p).symm⟩
+
+/-- The set of points in the affine span of a nonempty set of points
+is nonempty. -/
+lemma span_points_nonempty_of_nonempty {s : set P} :
+  s.nonempty → (span_points k V s).nonempty 
+| ⟨p, hp⟩ := ⟨p, mem_span_points k V p s hp⟩
+
+/-- Adding a point in the affine span and a vector in the spanning
+submodule produces a point in the affine span. -/
+lemma vadd_mem_span_points_of_mem_span_points_of_mem_vector_span {s : set P} {p : P} {v : V}
+    (hp : p ∈ span_points k V s) (hv : v ∈ vector_span k V s) : v +ᵥ p ∈ span_points k V s :=
+begin
+  rcases hp with ⟨p2, ⟨hp2, ⟨v2, ⟨hv2, hv2p⟩⟩⟩⟩,
+  rw [hv2p, vadd_assoc],
+  use [p2, hp2, v + v2, (vector_span k V s).add hv hv2, rfl]
+end
+
+/-- Subtracting two points in the affine span produces a vector in the
+spanning submodule. -/
+lemma vsub_mem_vector_span_of_mem_span_points_of_mem_span_points {s : set P} {p1 p2 : P}
+    (hp1 : p1 ∈ span_points k V s) (hp2 : p2 ∈ span_points k V s) :
+  p1 -ᵥ p2 ∈ vector_span k V s :=
+begin
+  rcases hp1 with ⟨p1a, ⟨hp1a, ⟨v1, ⟨hv1, hv1p⟩⟩⟩⟩,
+  rcases hp2 with ⟨p2a, ⟨hp2a, ⟨v2, ⟨hv2, hv2p⟩⟩⟩⟩,
+  rw [hv1p, hv2p, vsub_vadd_eq_vsub_sub V (v1 +ᵥ p1a), vadd_vsub_assoc, add_comm, add_sub_assoc],
+  have hv1v2 : v1 - v2 ∈ vector_span k V s,
+  { apply (vector_span k V s).add hv1,
+    rw ←neg_one_smul k v2,
+    exact (vector_span k V s).smul (-1 : k) hv2 },
+  refine (vector_span k V s).add _ hv1v2,
+  unfold vector_span,
+  change p1a -ᵥ p2a ∈ submodule.span k (vsub_set V s),
+  have hp1p2 : p1a -ᵥ p2a ∈ vsub_set V s, { use [p1a, hp1a, p2a, hp2a] },
+  have hp1p2s : vsub_set V s ⊆ submodule.span k (vsub_set V s) := submodule.subset_span,
+  apply set.mem_of_mem_of_subset hp1p2 hp1p2s
+end
+
+end affine_space
+
+open add_torsor affine_space
+
+/-- An `affine_subspace k V P` is a subset of an `affine_space k V P`
+that has an affine space structure induced by a corresponding subspace
+of the `module k V`. -/
+structure affine_subspace (k : Type*) (V : Type*) (P : Type*) [ring k] [add_comm_group V]
+    [module k V] [affine_space k V P] :=
+(carrier : set P)
+(direction : submodule k V)
+(nonempty : carrier.nonempty)
+(add : ∀ (p : P) (v : V), p ∈ carrier → v ∈ direction → v +ᵥ p ∈ carrier)
+(sub : ∀ (p1 p2 : P), p1 ∈ carrier → p2 ∈ carrier → p1 -ᵥ p2 ∈ direction)
```

Comment:
Here is an alternative approach. I'm not sure which one is better but I think that we should discuss the pros and cons of each approach.

1. Define `affine_subspace` using the following axioms:
```lean
structure affine_subspace (k : Type*) (V : Type*) (P : Type*) [ring k] [add_comm_group V]
    [module k V] [affine_space k V P] :=
(carrier : set P)
(add_mem' : ∀ p₁ p₂ p₃ ∈ carrier, (p₂ -ᵥ p₁) +ᵥ p₃ ∈ carrier)
(smul_mem' : ∀ (c : k) (p₁ p₂ ∈ carrier), c • (p₂ -ᵥ p₁) +ᵥ p₁ ∈ carrier)
```

These axioms are equivalent to yours with two differences:

1. I don't include `direction` in the data (though it can be reconstructed).
2. I don't require the carrier to be nonempty.

The main "pro" of this approach is that the affine subspaces defined this way form a complete lattice, and you can define `affine_span s = Inf {t : affine_subspace k V P | s ⊆ (t : set P)}` and have a `galois_connection` between `set P` and `affine_subspace k V P`.

As for `direction`, there are a few different possibilities:

1. define `direction' : with_bot (submodule k V)` to be `bot` for the empty affine space or the set of `p1 -ᵥ p2` otherwise.
2. define `direction s [nonempty (s : set P)] : submodule k V`;
3. define `direction s : submodule k V` to be the zero submodule for the empty affine space, otherwise as in 1.

Of course, you can add two of these definitions and prove that they agree (look at `polynomial.degree` vs `polynomial.nat_degree`).

Reply:
This does sounds pretty nice!

@jsm28, you're definitely allowed to include "sounds good, but I don't feel up to attempting this now" in your answer. :-)

Reply:
I think that may be my shorter answer, then.

As a longer answer, I suspect there might be use for both types eventually, but the one not requiring `nonempty` (i.e. sets of points containing all affine combinations) might best be called something other than `affine_subspace`. That one gives the abstract benefit of the lattice structure, but results in various other complications, and I'm less sure of having practical uses for it (but it may be hard to tell what complications arise with each approach without a lot more implemented that builds on affine subspaces).

Anything involving dimensions is likely to be more complicated with empty subspaces, because an empty space should have dimension -1, but dimensions are also naturally expected to be a nat, or a cardinal in the infinite-dimensional case, and to be the same as the dimension of the `submodule` that is the `direction`; dimension 0 is naturally equivalent to having a single point, so if you special-case empty subspaces as also having dimension 0 to keep the types convenient you lose that equivalence. Two subspaces should be parallel iff their directions are equal, but with empty subspaces that doesn't work if the direction has type `submodule` which would make empty spaces parallel to those with a single point. Without empty subspaces, two subspaces are equal iff they have nonempty intersection and the same direction. With empty subspaces, it's no longer the case that any subspace has an affine frame, and a subspace of dimension n only necessarily has a barycentric frame of n+1 points if your definition allows dimension -1 (and various definitions around barycentric coordinates would probably need special-casing for empty subspaces). Empty subspaces can't be converted to affine spaces on a subtype of points, whereas nonempty ones can.

As a general implementation principle, I hope that most definitions and proofs around affine subspaces and spans can do their main work through the corresponding definitions and results for submodules and spans in modules, rather than needing to repeat any logic that's already present in the module case.

I do know that for geometry I'll want types that build on top of nonempty subspaces by requiring them to have a given dimension, so that e.g. types `line k V P` or `plane k V P` can be defined to `affine_subspace_n_dim k V P 1` or `affine_subspace_n_dim k V P 2` or similar (and whenever a 0-dimensional subspace arises, it will probably be more convenient to convert it back to a point of type `P`).

I don't right now have any working code outside of this PR that builds on affine subspaces. (I do have code building on affine maps, at a fairly trivial level - constructing an `affine_map` from an isometry of Euclidean spaces.)

Reply:
Just to be clear:

* This was a suggestion, i.e. I won't insist at least until I'll need this for some good reason (most probably I won't have a reason to use this part of `mathlib` in the next several months); if I'll have a good reason, I'll come again and will be ready to do the refactor.
* I do not suggest proving facts about the "bad" empty subspace when it's not convenient. Just add `[nonempty (s : set P)]` or `(hs : (s : set P).nonempty)` to any `def`/`lemma`.